### PR TITLE
Fix refactor dispatcher

### PR DESCRIPTION
**Pull request by spion-h4**
Refactor the dispatcher. Split the logic for responding from the logic for dispatching and make sure that responses happen only AFTER the request lifecycle dispose has completed, rather than in parallel.

Tip: take a look at the test snapshot file, it will help you a lot with the review.

### CODE REVIEW

**Review by gphfour — COMMENTED** [review]
Looks great! The tests especially. A really nice approach that paints a clear image of how the dispatcher behaves. I suggest we just add some docs to it and we should be good to go. I will wait for the others to take a look too, and we should all be involved in the testing when we integrate it with the platform.

File: packages/backend/src/rpc/index.spec.ts [hunk 1]
```diff
@@ -0,0 +1,407 @@
...
+        params: {},
+        response: () => 'data',
+      });
+      expect(requestInfo.res.json).toHaveBeenCalledWith({
+        code: 200,
+        result: 'data',
+        error: null,
+        version: 2,
+      });
+      expect(requestInfo.res.status).toHaveBeenCalledWith(200);
```
[the remark this answers is not in the document]
Comment (spion-h4) [reply packages/backend/src/rpc/index.spec.ts:126]: I rewrote the tests to use runMock, which does this check for every test

File: packages/backend/src/rpc/index.spec.ts [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+    res?: Response;
+  },
+) {
+  let sendResponse = opts.response || (() => 'Hello World');
+  let mReq = opts.req || mockRequest(opts.method, opts.params);
+  let mRes = opts.res || mockResponse();
+
+  let disposed = false;
+  app.getSingleton(ServiceContextEvents).onContextDisposed(() => {
+    disposed = true;
```
Comment (spion-h4) [inline packages/backend/src/rpc/index.spec.ts:38]: might want to await a 1ms timeout promise here, just to make sure.

**Comment by spion-h4** [comment]
Close in favor of #81

**Pull request closed without merging**